discretize/base/base_mesh.py

```diff
@@ -1880,6 +1883,415 @@ def get_edge_inner_product(
+            \approx \sum_i \int_{S_i} \, \vec{u}_\perp \cdot \tau \vec{v}_\perp \, da
+            = \mathbf{u^T \, M_\tau \, w}
+
+        where the diagnostic properties on mesh faces (i.e. the model) are stored within an array of the form:
+
+        .. math::
+            \boldsymbol{\tau} = \begin{bmatrix} \boldsymbol{\tau_x} \\
+            \boldsymbol{\tau_y} \\ \boldsymbol{\tau_z} \end{bmatrix}
+
+        Examples
+        --------
+        Here we provide an example of face inner product surface matrix.
+        For simplicity, we will work on a 2 x 2 x 2 tensor mesh.
+        As seen below, we begin by constructing and imaging the basic
+        face inner product surface matrix.
+
+        >>> from discretize import TensorMesh
+        >>> import matplotlib.pyplot as plt
+        >>> import numpy as np
+        >>> import matplotlib as mpl
+
+        >>> h = np.ones(2)
+        >>> mesh = TensorMesh([h, h, h])
+        >>> Mf = mesh.get_face_inner_product_surface()
+
+        >>> fig = plt.figure(figsize=(6, 6))
+        >>> ax = fig.add_subplot(111)
+        >>> ax.imshow(Mf.todense())
+        >>> ax.set_title('Basic Face Inner Product Surface Matrix', fontsize=18)
+        >>> plt.show()
+
+        Next, we consider the case where the physical properties
+        are defined by diagnostic properties on mesh faces. For the isotropic case,
+        we show the physical property tensor for a single cell.
+
+        Define the diagnostic property values for x, y and z faces.
+
+        >>> tau_x, tau_y, tau_z = 3, 2, 1
+
+        Here construct and image the face inner product surface matrix for the isotropic case.
+        Spy plots are used to demonstrate the sparsity of the inner product surface matrices.
+
+        >>> tau = np.r_[tau_x * mesh.n_faces_x, tau_y * mesh.n_faces_y, tau_z * mesh.n_faces_z]
+        >>> M = mesh.get_face_inner_product_surface(tau)
+
+        Then plot the sparse representation,
+
+        >>> fig = plt.figure(figsize=(4, 4))
+        >>> ax1 = fig.add_subplot(111)
+        >>> ax1.spy(M, ms=5)
+        >>> ax1.set_title("M (isotropic)", fontsize=16)
+        >>> plt.show()
+        """
+        try:
+            if model is None:
+                model = np.ones(self.nF)
+
+            if invert_model:
+                model = 1.0 / model
+
+            if is_scalar(model):
+                model = model * np.ones(self.nF)
+
+            # Isotropic case only
+            if model.size == self.nF:
+                Aprop = self.face_areas * mkvc(model)
+                M = sdiag(Aprop)
+            else:
+                raise Exception(
+                    "Unexpected shape of tensor: {}".format(model.shape),
+                    "Must be scalar or have length equal to total number of faces.",
+                )
+        except NotImplementedError:
+            raise NotImplementedError(
+                f"get_face_inner_product_surface not implemented for {type(self)}"
+            )
```

Comment:
There are a lot of things being tried in this statement here, and it's slightly unclear to someone reading the code what statements you're actually testing for exceptions. I'd recommend as follows:

```suggestion
        try:
            areas = self.face_areas
        except NotImplementedError:
            raise NotImplementedError(
                f"get_face_inner_product_surface not implemented for {type(self)}"
            )

        if model is None:
            model = np.ones(self.nF)

        if invert_model:
            model = 1.0 / model

        if is_scalar(model):
            model = model * np.ones(self.nF)

        # Isotropic case only
        if model.size != self.n_faces:
            raise ValueError(
                f"Unexpected shape of tensor: {model.shape}. "
                "Must be scalar or have length equal to total number of faces.",
            )

        Aprop = areas * mkvc(model)
        M = sdiag(Aprop)
```

This makes it fail faster if the mesh hasn't implemented a `face_areas` property yet (and it makes it clear in the code that that is what you are testing whether or not it is implemented). It also separates the check for the size of the model from constructing the matrix. (It does the check first then proceeds if it is all good.)